design/pipeline-caching.md

```diff
@@ -0,0 +1,157 @@
+# Pipeline Caching
+
+*** Status: Work in Progress ***
+
+We have observed the need for Azure Pipelines to implement a dependency caching mechanism that allows the outputs of steps within a pipeline to be optimized (or even skipped) if a suitable cached version of the outputs of those steps already exists. Caching can be effective provided the cost of determining a cache hit and acquiring the contents of the cache is cheaper than the cost of producing the output again from scratch.
+
+## Proposal
+
+This proposal contains a number of interconnected elements. We have attempted to layer them in this document from the most primitive building blocks up to simplified YAML syntax to make caching "just work" in the majority of cases. Our approach to building a caching mechanism for Azure Pipelines will be to get the fundamental building blocks right and prove we can increase build performance and then evolve the YAML syntax once we know exactly what that syntax will need to describe.
+
+### Restore and Save Cache Tasks
+
+The basic functions of Pipeline Cache will be implemented by two tasks. The ```RestoreCache``` task and the ```SaveCache``` task. Both tasks will specify a collection of files and environment variables which can be used as a key to lookup the cache and identify a hit or a miss. Each task will also take a collection of paths which will be fetched from or stored in the cache associated with that key.
+
+Consider the following Azure Pipelines pipeline defined in YAML:
+
+```yaml
+steps:
+- script: yarn
+- script: yarn build
+```
+
+In this example we are working on a React-based application which was bootstrapped using ```create-react-app```. The output of this command is a directory with about ~28,000 files including the ```node_modules``` directory. On a local developer machine the ```yarn``` command will trigger the installation of any missing packages. Typically this will execute quickly because all packages already exist on the developer machine. However in cloud-based CI environments where build machines are recycled after every build there is no persistent state which means that Yarn must download all packages from scratch.
+
+Applying caching to this pipeline would involve adding a restore cache step and save cache step before and after the ```-script: yarn``` step.
+
+```yaml
+steps:
+- task: RestoreCache@0
+  inputs:
+    key: |
+      package.json
+      yarn.lock
+    paths: |
```

Comment:
Will it be possible to define full paths on the system? For example when using Maven the `~/.m2` needs to be cached. Or in this case is the idea for caching the Maven repo to be part of the Maven task?

Reply:
This is a problem we are grappling with at the moment. You’ll certainly be able to cache the directory but when restoring you might need to tell it where it is going back to since we don’t capture the full path in the cache manifest today. We need to think through this some more because it impacts ease of use on restore but also whether we can place multiple unrelated trees of content in the same cache (thanks for raising it).

As for task integration, we envisage that once we get these building blocks created that tasks will be able to integrate directly as well.

In YAML usage scenarios we will probably leverage the use: statement which would simply inject tasks with the right paths preconfigured.

We want to get the building blocks right first though.

Reply:
Was trying some things out and I just realised that I can tell maven to use a folder relative to the working directory for `.m2` when invoking the commands so that is OK. The can be done by setting `-Dmaven.repo.local=.m2` for the maven options. You might need to consider this as an example for maven